### ISSUE TITLE
Ability to easily override ROM 

### ISSUE THREAD

**Issue by paulscottrobson** [lgblgblgb/xemu#69]
Hi. I'm (hopefully) going to be writing the BASIC ROM for the Mega 65, well possibly two. What I would like is an easy way of booting xemu with an alternate c65-system.rom. The simplest method , modifying the SDCard seems to have problems beyond speed, and things like the cache load option I can't make work, this might be me (!). At present I'm trapping the kickstart with PC=$A564 (the branch test after the ROM has loaded) and overwriting it there, but this isn't a particularly good solution, even though it does seem to work. It would be handy if one could simply override the system.rom as provided in the SDCard.

**Comment by lgblgblgb** [comment]
It's a bit hard thing to do, since - unlike the C65 emulator - Mega65 ROM is actually loaded from SD-card by KickStart, and I can't override what KickStart does, since it's the "real" one, from the Mega65 project. So the ROM loading is not done by Xemu at all. What you can try to do: use the -loadrom option of Xemu/M65. It does a "preload" process into the memory space, before KickStart even runs. I'm not sure what would happen, in theory KickStart has some mechanism to check if there is a valid "ROM" in the memory already, but I never tried this one ... Really it's up to the KickStart ... Second more complicated option: compile & modify a KickStart from source itself, and make Xemu/M65 to use that KickStart (-kickup option). But please note, that Xemu (btw, use the "dev" branch if compiled from source ...) is kinda old, maybe not compatible with KickStart too recent inside the mega65-core project tree ...

So in nutshell: I would love to help, but it's not so easy to do, since it's kinda scope of the system software (KS) not Xemu too much.

**Comment by lgblgblgb** [comment]
Btw, if you're using UNIX like OS, it's easy to find tools which can write FAT32 images as well from command line, (like mtools on Linux and such) so you can write a shell script which copies your ROM to the SD-card image and launch Xemu/M65. So then you need to run a single command to have Xemu/M65 "booting" with your ROM, the same as before, just you run another "executable" (a shell script). I have no idea about Windows though, I never used Windows, I only made Xemu to be able to run on it (not so much even checked if it works, to be honest, only a few times). Yes, I admit, it's an ugly solution. But feel free to suggest some solutions, but if you read my previous comment you see the problem why it's so hard to implement this feature, that's the reason I suggested outside-of-xemu solution (this comment) or KS-level solution (previous comment).

**Comment by paulscottrobson** [comment]
Yes, having written a few emulators myself it would actually be a hack to the system which is basically how I've done it ; jamming a brute force load in at a specific point in the boot ROM. Often you do a similar thing with loading off cassettes. I couldn't get the --loadrom to work, but I'll give it another go. I'm an Arch user, can't abide Windows. Not that I could use it anyway as the whole build install on A7 board stuff is done on Linux.

**Comment by lgblgblgb** [comment]
I see your point, however there is a little problem here. This theory you mentioned (and I also used them on many emulators of mine actually as well) based on well-established ROM contents of those machines, well-known and unchanged since decades or so. In case Mega65 this actual ROM is the kickstart (KS for short) which is from actual source code from the mega65-core project and always changes on source code level during the still on-going development. So it's not possible to just "patch it" at least not at binary level, because of the mentioned reasons. SURELY, at source code level it can be patched, I admit that, but then it would introduce a divergence between the real KS. The point of Xemu/M65 is to be able to use the unmodified every changing (at least currently) development material of the M65 project itself, also the origin of this emulator was to help to development of the project itself. One of the paths I can see realistic in the future, is to add FAT32 handling into Xemu itself. It's often a wish from users as well, that they don't know how to install their ROMs, D81s etc on the sd-card image. Xemu would be able to open the SD-card image and place the needed files there on user's request, and then only start to execute the emulation of the machine for real, so it will see the files the user want, basically. With this theory there can be a command line switch to "import" files on the SD-card image by Xemu itself which then can be used for purposes you want too, I guess ...

The -loadrom switch (note, that only one dash before the option btw) loads the ROM image to the memory. Maybe that does not help you, if KS loads another from the SD-card image, thus overwriting the one in the memory. You may try however to use an SD-card image file which does not have the ROM image at all, I haven't tried out what happens, KS would "panic" not having ROM, or using the in-memory one (I have a faint memory that it may work, but then it writes back the ROM content into the SD-card image file to help user, but actually it make things worse with your kinda specific situation).

**Comment by lgblgblgb** [comment]
Btw, I more or less suggested this in a previous comment, that you can "fed" a custom KickStart for Xemu which skips loading ROM from the SD controller, so by using -kickup option and -loadrom at once, with a modified KS, it would probably be able to do what you want, since then you can modify the KS as much as you want (now probably just to skip loading the ROM and leave memory as is, to allow -loadrom has effect ...). Maybe the naming scheme can give you some confusion here: -loadrom loads the "ROM image" ie C65 ROM etc, -kickup loads the KS, that's the software specific for M65 (does not exist on C65) which loads the C65 ROM as well, among other things, like the Hypervisor, HDOS (Hypervisor DOS, which can read FAT32), etc.

**Comment by lgblgblgb** [comment]
Ok, the third post since your last one, hopefully it's OK, I really would like to help. If you want to write a custom ROM and it wanted to be C65 compatible (ie, not using M65-only feature, since M65 itself "should be" compatible with C65 ...) then you can use Xemu's C65 emulator instead of M65. The benefit, that you can forget any Kickstart/etc thing of M65 and it works like a regular 8 bit machine, it has ROM, which is not loaded, etc etc. There you can easily specify ROM without confusion by KS or anything then ... But you probably want to use M65-specific things as well, then this idea won't work, of course ....

**Comment by paulscottrobson** [comment]
The Kickstart, or Hippo, or whatever ROM isn't changing that much. At some point it's going to call the code to load MEGA65.ROM. All I've done is to hack the emulator (not in your copy) so that once it's loaded that it overwrites it with my WIP ROM ; it just tests the PC in a specific state, it's a few lines. It's actually preferable because that way I'm not changing the KS ROM. It's simpler than messing about with SD Card images as well :) It's only a small part ; most of the development is done on a custom built emulator, M65 specific stuff is done on your emulator and obviously the NexysA7 board for the real thing which has monitor_load to load in kickstart and BASIC ROMs directly anyway.

Don't worry about it, I'm quite happy with my solution. All it means is that when the KS changes I've got to dig out the load point again, which takes a few minutes tops.

**Comment by lgblgblgb** [comment]
Well, since I am a developer of the project, I know it changes (but I mean from the point of binary patching what you proposed with examples in your comment), that's the reason I wrote, not because of other :) And well, yes, in theory it's called Hippo now, or whatever, but the old habits, I still name it KS, when I saw the the code forming.

It's kinda fun to write an emulator for yourself, the hard part when users have wishes, and I don't write about you now, but in general with much more simple things, which can be complicated because of how M65 works in contrast of many old systems, ie it has SD-card, it loads things from there etc (basically at FPGA level it has no ROM at all, OK, surely the bitstream for the FPGA contains the KS code, but it's in the "hypervisor RAM" which is for real a block RAM inside FPGA, and can be even overwritten, not even talking about the C65 ROM which is not part of the bitstream either, while a real C65 would contain its ROM as ROM, and no magic "hypervisor" or anything like that). One good example is the "auto load" feature. For a classic system, ROM patching or at least watching PC of the CPU after startup / etc, can be the way. However for C65 there are many different ROMs. So this is harder to do, and in this case basically I try to detect the start of the keyboard scanning from the "emulated hardware side" and I emulate the keypresses. It's kinda not ideal, indeed. Here, we see even the original ROM is question of "which?", not so much a single well-known one. KS is even more "problematic" since it's under development still right now ... Sorry for the long post, just I wanted express my point.

But, as usual, I'm always open to hear people's opinion. Basically, writing emulators is just a hobby of mine, and I can't even say, that I'm so professional to do so, I just have the desire to do it what I can do.

Happy to hear that you solved the problem in a "hacky way", honestly, I often has the same kind of tricks especially for myself. That's find if you do it for yourself, the harder part, when you want to present a feature to the "public", I think :) But I also think, that your original problem is kinda a real one and connected to many similar wishes, which "should be" solved somehow after all in the "official" version of Xemu/M65 as well. For example because of things I've already written, people even have problem to "install" the ROM image on the SD-card-image _once_ even if they have no intent to develop ROM or anything advanced you do. That's why I was so in writing about features (like SD-card image modification by Xemu itself) you may found "too heavy" for your particular problem at least.

Sorry if it's hard to follow my long writings and thoughts, honestly I always found it hard to express myself well enough and especially in a short form, and well, my not so perfect English does not help on this either ...

Keep up your work, and surely, just drop me an issue, or even private PM, or whatever if I can do anything to help, or just hear some opinion and suggestion! Thanks.

**Comment by lgblgblgb** [comment]
After all, I could find a simple but still somewhat elegant way to this without too much ugly hack. See the comment of the commit ba27df64be8f5f30ce609816971dbc718b744f3a

It's in the "dev" branch as usual, the master one is always kinda old ...

**Comment by lgblgblgb** [comment]
Please let me know, if the current version of the "dev" branch (one extra commit since then) solves your particular need, or you have any problem with it. It should work in theory at least, and it allow many other user scenarios, that's why the split between separated -forcerom and -loadrom which both of needed by you in case of your workflow at least.

**Comment by lgblgblgb** [comment]
Now I've tested it with a custom ROM, from the original C64 (!) ROM [not C65] with 0xA000 bytes of zeroes, then the BASIC ROM, 4K of zeroes, then the character ROM, then the KERNAL, then 64K of zeroes, thus 128K in size (thus the expected layout of the ROM area of even C65 not only M65). By specifying that ROM with -loadrom and using -forcerom, Xemu/M65 managed to use it without problem (even though the SD card image has the C65 "normal" rom for M65), as basically a "C64" then :) That is, at least for now, I think this function is well implemented and well tested enough.

**Comment by paulscottrobson** [comment]
I was using the master, which doesn't have forcerom - it now works perfectly. Thank you :)

**Comment by lgblgblgb** [comment]
Actually, master seriously lacks behind the "current" version which is the dev branch, my fault, I know. At least I did some sync to the master not long time ago, but in general it's better idea to use the "dev" branch, especially if you are an advanced user who is not afraid of some problems ;-P But yeah, that's not so normal to organize the work this way, I know ...
